### DISCUSSION TITLE
Authentication headers in Budibase REST API Integration

### DISCUSSION THREAD

**Discussion by jeremiah-blessing** [Budibase/budibase#2685] [Help] [answered]
We are trying to integrate Budibase into our internal tool chain. Our database has critical information which we can't allow Budibase to directly access with DB credentials. We have APIs with JWT authentication to our servers which we need to integrate. Right now, I am not able to find any documentation regarding adding authentication headers. We will also need to have a fetch request to log in and get JWT tokens and place them in authentication headers. 

Please let me know if this can be solved by any methods.

**Comment by mjashanks** [comment, marked as answer]
Hi Jeremiah.

This should be possible with Budibase. Steps below, with video. I appreciate that there is a bit of work in this and that you would have to set up your Authentication call for every screen.

Soon, we will be looking into improving the REST API connector. Would be good to get your feedback on this experience!

Below is a rather contrived example that should help. What I am doing below is
1. Fetching the `budibase/budibase` repository from the GitHub API
2. Using a field from that response to fetch issues for that repository

In reality, I don't need to to the first call in my example, but am doing so because it fits your example!

API Setup
1. Setup the REST API connection, with a root URL of `https://api.github.com`
2. Added a query ("Get Repo") to `GET /repos/budibase/budibase`, and saved it.
3. Added a query ("List Issues") to `GET /repos/budibase/{{ repo }}/issues`. Note the `{{ repo }}` in the path...
4. Added a `repo` parameter and give it a default value, so I can run this query now to test it. This parameter will be used inside the path, in step 3 above. **Note: you can add a HEADER and use a parameter inside it... e.g. `Authorization: Bearer {{ token }}`**

Now, in the design section
1. Added a new data provider, reads from "Get Repo"
2. Added a `Repeater` below
3. Added a new data provider, reads from "List Issues"
4. Click the little Cog icon, which allows me to input the `repo` parameter
5. Set the `repo` parameter to `{{ repos.Get budibase repo.name }}` - i.e. the name of the repo, returned by the "Get Repo" call
6. Added a Table component, that displays my issues. 

https://user-images.githubusercontent.com/3524181/134153242-71f6519c-75ca-43e0-886e-2af6e3b06d22.mp4

**Comment by jeremiah-blessing** [comment]
Hi @mjashanks ,

Sorry for the delayed response. Thank you for your crystal clear step-by-step explanations with a Video. I followed every step and I am able to achieve what we had in mind. 

```
┌────────────────────────────────────────────────────────────────┐
│ Sign in - POST /login                                          │
│ ┌──────────────────────────────┐                               │
│ │ Data : jwt_token, user_name  │                               │
│ └──────────────────────────────┘                               │
│                                                                │
│ ┌────────────────────────────────────────────────────────────┐ │
│ │ Get all Products - GET /products with auth header          │ │
│ │ 'jwt_token' from parent provider                           │ │
│ │ ┌───────────────────────────────────┐                      │ │
│ │ │ Data : [name, price, quantity,..] │                      │ │
│ │ └───────────────────────────────────┘                      │ │
│ │                                                            │ │
│ │ Shown all products in table                                │ │
│ │                                                            │ │
│ └────────────────────────────────────────────────────────────┘ │
│                                                                │
└────────────────────────────────────────────────────────────────┘


```

One downside that we are facing with this setup is, every time a user refreshes the page, a new jwt token is being generated. We will look into it and if possible we will contribute to the REST integration to have authentication setup prebuilt. 

Thank you :)

**Comment by chrisshaw** [comment]
@jeremiah-blessing this is a great question and @mjashanks this was a great answer. Pardon the diligence as we go down the same path:

- @jeremiah-blessing If I may ask, how are you setting up authentication? What's the user experience? We're using Auth0 in Budibase's OIDC. I'm trying to think how this would work with Auth0's auth code flow.
- @mjashanks Do you have a sense of roadmap priority on enhancing the REST connector?

**Reply by mjashanks** [reply]
Hey @chrisshaw. It's not officially scheduled, so I can't make promises. It won't be next month. I would like to see it this year, but that goes for a lot of things 😀

It's very important. It would help us if you could tell us what you think the limitations of the current implementation are. Thanks!

**Reply by chrisshaw** [reply]
Sure thing, I'll open a separate discussion.

**Reply by jeremiah-blessing** [reply]
Hi @chrisshaw, We have a seperate authentication server and data servers. Data servers check and validate token generated by authentication servers.

**Comment by SantoshSrinivas79** [comment]
@jeremiah-blessing I'm trying to do something similar but stuck with the JWT token. Would you be able to share how you achieved the following part:

> Right now, I am not able to find any documentation regarding adding authentication headers. We will also need to have a fetch request to log in and get JWT tokens and place them in authentication headers.

Appreciate any help. Thanks in advance!

**Reply by mjashanks** [reply]
@SantoshSrinivas79 in the next few weeks, we will be releasing a new REST API connector, that allow for this. Just letting you know.

**Reply by jeremiah-blessing** [reply]
@SantoshSrinivas79 Can you please explain your JWT auth flow?

In my case, I set the JWT token in the header as `Authorization`.

**Reply by SantoshSrinivas79** [reply]
@jeremiah-blessing ... I'll just tweak mine to mimic your process for now (till the new RESTI API connector is released). Would really appreciate a screen recording if possible.

> Right now, I am not able to find any documentation regarding adding authentication headers. We will also need to have a fetch request to log in and get JWT tokens and place them in authentication headers.
